R/list_plot.R

```diff
@@ -53,6 +77,18 @@ list_plot_xy <- function(df, x, y, .fun = cont_cont, ...) {
       k <- k + 1
     }
   }
+  clx <- col_labels(x)
+  cly <- col_labels(y)
+  if(length(x)==1) {
+    outn <- cly
+  }
+  if(length(y)==1) {
+    outn <- clx
+  }
+  if(length(y) > 1 && length(x) > 1) {
+    outn <- lapply(clx, function(xi) paste0(cly, "v", xi))
+  }
```

Comment:
If either `x` or `y` is empty, `outn` is undefined (same applies to `list_plot_yx`):

```R
list_plot_xy(
  pmplots_data_id(),
  x = c("STUDYc//Study", "CPc//Child-Pugh"),
  y = c(),
  .fun = cont_cat
)
#> Error in list_plot_xy(pmplots_data_id(), x = c("STUDYc//Study", "CPc//Child-Pugh"),  :
#>   object 'outn' not found
```

From a user standpoint, I realize passing an empty `x` or `y` is not valid or useful.  Also, `list_plot_x` and `list_plot_y` will error in other ways if `y` or `x`, respectively, is empty.

Still I think it'd be good to restructure these conditions to avoid the possibility of an undefined variable.

<hr>

For `list_plot_xy` and `list_plot_yx`, another thing I wondered about is the `length(x) == 1 && length(y) == 1` scenario. Both functions set the names to the `x` element in that case. I suppose it doesn't really matter, because it's not an interesting case, but I wonder whether it'd be better to just stick with the "&#x2026;v&#x2026;" naming in that situation.

<hr>

So, all together, I'm suggesting something like this patch:

<details>
<summary>patch</summary>

```diff
diff --git a/R/list_plot.R b/R/list_plot.R
index 03de468..89cbbf3 100644
--- a/R/list_plot.R
+++ b/R/list_plot.R
@@ -25,6 +25,7 @@
 #' @export
 #' @rdname list_plot
 list_plot_x <- function(df, x, y, .fun = cont_cont, ...) {
+  stopifnot(length(x) > 0, length(y) > 0)
   out <- vector(mode = "list", length = length(x))
   for(i in seq_along(x)) {
     out[[i]] <- .fun(df = df, x = x[i], y = y, ...)
@@ -48,6 +49,7 @@ list_plot_x <- function(df, x, y, .fun = cont_cont, ...) {
 ##' @export
 ##' @rdname list_plot
 list_plot_y <- function(df, x, y, .fun = cont_cont, ...) {
+  stopifnot(length(x) > 0, length(y) > 0)
   out <- vector(mode = "list", length = length(y))
   for(i in seq_along(y)) {
     out[[i]] <- .fun(df = df, x = x, y = y[i], ...)
@@ -69,6 +71,7 @@ list_plot_y <- function(df, x, y, .fun = cont_cont, ...) {
 ##' @export
 ##' @rdname list_plot
 list_plot_xy <- function(df, x, y, .fun = cont_cont, ...) {
+  stopifnot(length(x) > 0, length(y) > 0)
   out <- vector(mode = "list", length = (length(x)*length(y)))
   k <- 1
   for(i in seq_along(x)) {
@@ -79,13 +82,11 @@ list_plot_xy <- function(df, x, y, .fun = cont_cont, ...) {
   }
   clx <- col_labels(x)
   cly <- col_labels(y)
-  if(length(x)==1) {
-    outn <- cly
-  }
-  if(length(y)==1) {
+  if (length(y) == 1 && length(x) > 1) {
     outn <- clx
-  }
-  if(length(y) > 1 && length(x) > 1) {
+  } else if (length(x) == 1 && length(y) > 1) {
+    outn <- cly
+  } else {
     outn <- lapply(clx, function(xi) paste0(cly, "v", xi))
   }
   names(out) <- unlist(outn, use.names=FALSE)
@@ -95,6 +96,7 @@ list_plot_xy <- function(df, x, y, .fun = cont_cont, ...) {
 ##' @export
 ##' @rdname list_plot
 list_plot_yx <- function(df, x, y, .fun = cont_cont, ...) {
+  stopifnot(length(x) > 0, length(y) > 0)
   out <- vector(mode = "list", length = (length(x)*length(y)))
   k <- 1
   for(j in seq_along(y)) {
@@ -105,13 +107,11 @@ list_plot_yx <- function(df, x, y, .fun = cont_cont, ...) {
   }
   clx <- col_labels(x)
   cly <- col_labels(y)
-  if(length(x)==1) {
-    outn <- cly
-  }
-  if(length(y)==1) {
+  if (length(y) == 1 && length(x) > 1) {
     outn <- clx
-  }
-  if(length(y) > 1 && length(x) > 1) {
+  } else if (length(x) == 1 && length(y) > 1) {
+    outn <- cly
+  } else {
     outn <- lapply(cly, function(yi) paste0(yi, "v", clx))
   }
   names(out) <- unlist(outn, use.names=FALSE)
diff --git a/tests/testthat/test-list_plot.R b/tests/testthat/test-list_plot.R
index 545456b..660226c 100644
--- a/tests/testthat/test-list_plot.R
+++ b/tests/testthat/test-list_plot.R
@@ -57,6 +57,11 @@ test_that("list_plot_xy [PMP-TEST-018]", {
   expect_equal(names(p)[4:6], c("AAGvCRCL", "SCRvALB", "AAGvALB"))
 })

+test_that("list_plot_xy, single x and y", {
+  p <- list_plot_xy(df, x[1], y[1])
+  expect_identical(names(p), "SCRvWT")
+})
+
 test_that("list_plot_yx", {
   p <- list_plot_yx(df, x, y)
   expect_equal(length(p), length(y)*length(x))
@@ -64,3 +69,8 @@ test_that("list_plot_yx", {
   expect_equal(names(p)[1:3], c("SCRvWT", "SCRvCRCL", "SCRvALB"))
   expect_equal(names(p)[4:6], c("AAGvWT", "AAGvCRCL", "AAGvALB"))
 })
+
+test_that("list_plot_yx, single x and y", {
+  p <- list_plot_yx(df, x[1], y[1])
+  expect_identical(names(p), "SCRvWT")
+})
```

</details>

What do you think? (Again, I don't think any of this matters for normal usage; the undefined variable just got me looking into it more closely.)

Reply:
This is all great. I re-did this a couple of times and left it with an ambiguous feeling about what to do. Thanks for taking the time to think / work through this code and for the patch. I applied everything and I trust it the most at this point.